### PR TITLE
Add a placeholder call symbol for jProfileValue

### PR DESCRIPTION
**Pull request by r30shah**
Add following placeholder call symbol to CommonNonHelperSymbol enum for jProfilingValue.

1. jProfileValue : Symbol for regular value profiling which does not need NULLCHK
2. jProfileValueWithNullCHK : Symbol for value profiling that requires NULLCHK

Signed-off-by: Rahil Shah <rahil@ca.ibm.com>

### CODE REVIEW

**Comment by r30shah** [comment]
@andrewcraik  I have added two placeholder symbols (One for regular profiling and one that needs NULLCHK) as per suggestion in https://github.com/eclipse/openj9/pull/6585#discussion_r308331882 
Can I get a review on this change?

**Comment by andrewcraik** [comment]
You probably want a findOrCreate to get the symbol reference. We also need to consider the aliasing - none may be right since the call has no side-effects.

File: compiler/compile/OMRSymbolReferenceTable.cpp [hunk 1]
```diff
@@ -2013,6 +2013,20 @@ TR_BitVector *OMR::SymbolReferenceTable::getSharedAliases(TR::SymbolReference *s
    return NULL;
    }
 
+
+TR::SymbolReference *
```
Comment (r30shah) [inline compiler/compile/OMRSymbolReferenceTable.cpp]: I have added the findOrCreate for this type of symbols. These symbols are similar to InlinedByCodeGen symbols which will be inlined by the codegen (https://github.com/eclipse/omr/blob/8acafad8398cc9a596ba7d4f0821e18853711803/compiler/compile/OMRSymbolReferenceTable.cpp#L766) . Delta here would be that these will be lowered to trees by optimizer.

Comment (andrewcraik) [reply compiler/compile/OMRSymbolReferenceTable.cpp]: passing in the index opens this API to misuse - the convention is one per special helper so I think that would make more sense and keep the code more terse...

Comment (r30shah) [reply compiler/compile/OMRSymbolReferenceTable.cpp]: I had `FATAL_ASSERT` to prevent misuse. But following the convention here, separated out the functions for both symRef in  https://github.com/eclipse/omr/pull/4172/commits/619a159a143a9124a51498c9b209f18ccc5d4029
Let me know changes are fine. Will squash the commits.

Comment (andrewcraik) [reply compiler/compile/OMRSymbolReferenceTable.cpp]: looks good to me

**Comment by andrewcraik** [comment]
@genie-omr build all

**Comment by andrewcraik** [comment]
over to @vijaysun-omr as assignee

File: compiler/compile/OMRSymbolReferenceTable.cpp (hunk 1 above)
Comment (r30shah) [reply compiler/compile/OMRSymbolReferenceTable.cpp]: Thanks a lot.. once builds are finished, will squash commits.

**Comment by vijaysun-omr** [comment]
What is the "placeholder" aspect here ? i.e. why is that word in:

TR::SymbolReference * findOrCreateJProfileValuePlaceHolderSymbolRef();
TR::SymbolReference * findOrCreateJProfileValuePlaceHolderWithNullCHKSymbolRef();

**Comment by r30shah** [comment]
@vijaysun-omr, These symbol references are used to represent a value profiling site identified by other optimization. We will have following placeholder calls added by other optimizations for the values they decide to profile.
‘’’
call jProfileValueSymbol
       <value to profile>
       <table address>
‘’’
We lower this place holder calls to trees in ‘jProfiingValue’ phase. 
Previously we were using a three child call to actual jProfile32BitValue/64BitValue of the following form to represent placeholders.
‘’’
call jProfile32BitValue
      <value to profile>
      <table address>
      iconst <some value>
‘’’ 
In jProfilingValue phase we were inspecting childs of calls to differentiate actual helper calls and placeholder calls.
Now if we are profiling vft load if object under instanceOf/checkcast where a object can be NULL so we would need a null test to avoid dereferencing NULL in profiling trees. To represent these value profiling site s.t it can add null test, we need second symref jProfileValueWithNullCHKSymbol.

So coming back to your question, as calls to these symbols represent just the value profiling site and do not reach code gen, they are place holders.

**Comment by r30shah** [comment]
@vijaysun-omr, These symbol references are used to represent a value profiling site identified by other optimization. We will have following placeholder calls added by other optimizations for the values they decide to profile.
```
call jProfileValueSymbol
      <value to profile>
      <table address>
```
We lower this place holder calls to trees in ‘jProfiingValue’ phase. 
Previously we were using a three child call to actual jProfile32BitValue/64BitValue of the following form to represent placeholders.
```
call jProfile32BitValue
     <value to profile>
     <table address>
     iconst <some value>
```
In jProfilingValue phase we were inspecting childs of calls to differentiate actual helper calls and placeholder calls.
Now if we are profiling vft load if object under instanceOf/checkcast where a object can be NULL so we would need a null test to avoid dereferencing NULL in profiling trees. To represent these value profiling site s.t it can add null test, we need second symref jProfileValueWithNullCHKSymbol.

So coming back to question, These symbols represent a placeholder calls for value profiling site which should not reach to codegen,

**Comment by vijaysun-omr** [comment]
Thanks, when you say : "In jProfilingValue phase we were inspecting childs of calls to differentiate actual helper calls and placeholder calls."....by "actual helper calls" you mean helper calls unrelated to JProfiling that are not lowered (and therefore are'nt considered placeholders) ?

**Comment by andrewcraik** [comment]
so the slow path of the JProfiling code - where we need to update the table of values - uses a helper call. Those helpers are real helpers. Call to these helpers are added to the trees in the process of replacing the 'place holder's with the tree implementation of the profiling. To avoid confusion @r30shah has created these non-helpers to track values we wish to profile. jProfilingValue will then process these calls to insert the profiling implementation which will include a call to the table modification helper.

TL;DR; the two new symbols here do not survive to codegen, the lowering of these symbols involves adding a call to an existing helper symbol to update the table on a slow path which is what were referred to as actual helper calls.

**Comment by vijaysun-omr** [comment]
Okay thanks. This looks like a pretty safe infrastructural change. Merging.
